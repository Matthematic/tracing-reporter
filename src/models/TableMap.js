import Test from './Test';
import Table from './TestTable';
import TestTable from './TestTable';
const _ = require('underscore');


class TableMap {
    constructor(json) {
      this.tables = [];

      if (json && Array.isArray(json.tables)) {
        json.tables.forEach(x => {
          if (Array.isArray(x.tests)) {
            const table = new Table(x.id);
            x.tests.forEach(y => {
              table.add(new Test(y))
            });
            this.tables.push(table);
          }
        });
      }
    }

    add(thing) {
      if (thing instanceof Test) {
        if (!this.tables.find(t => t.id === thing.id)) {
            this.tables.push(new TestTable(thing.id).add(thing));
        }
        else {
            this.tables.find(t => t.id === thing.id).add(thing);
        }
      } else if (thing instanceof TestTable) {
        this.tables.push(thing);
      }
      else if (Array.isArray(thing)) {
        thing.forEach(t => this.add(t));
      }
      else {
          console.log("couldnt add", thing);
      }
      return this;
    }

    getTests() {
        return _.flatten(this.tables.map(t => t.getTests()));
    }

    /**
     * Splits tests that have combined issues into separate tests e.g. [TRACE-1000, TRACE-1001] becomes [TRACE-1000] [TRACE-1001]
     * @param {array} data 
     */
    getTestsSplitByIssue() {
      let dataCopy = [];
      _.flatten(this.tables.map(t => t.getTests())).forEach(row => {
        if (row.issues.length < 2) {
          dataCopy.push(row);
        }
        else {
          for (let i = 0; i < row.issues.length; ++i) {
            const rowCopy = Object.assign({}, row);
            rowCopy.issues = row.issues[i]; 
            dataCopy.push(new Test({...rowCopy}));
          }
        }
      });
  
      return dataCopy;
    }

    getTables() {
      return this.tables;
    }

    sort(key = 'id', direction = 'ascending') {
      this.tables.forEach(t => t.sort(key, direction)); // Sort each table
      this.tables = this.tables.sort((t1, t2) => Number(t1.id) < Number(t2.id) ? 1 : -1) // sort the list of tables in DESCENDING ID order
      return this;
    }

    count() {
      return this.tables.length;
    }

    print() {
        this.tables.forEach((t) => {
            t.print();
        })
    }

}

export default TableMap;