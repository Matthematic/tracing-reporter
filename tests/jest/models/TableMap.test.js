import TableMap from "../../../src/models/TableMap";
import Test from "../../../src/models/Test";
import TestTable from "../../../src/models/TestTable";

const unitTest = new Test({ id: '1', name: 'Unit Test', link: '"../testUnit.js#L3",', issues: ['TEST-123', 'TEST2-123'], shortLink: 'testUnit.js' , type: 'Unit' })
const grayboxTest = new Test({ id: '2', name: 'Graybox Test', link: '"../testGraybox.js#L3",', issues: ['TEST-223', 'TEST2-223'], shortLink: 'testGraybox.js' , type: 'Graybox' })
const blackboxTest = new Test({ id: '3', name: 'Blackbox Test', link: '"../testBlackbox.js#L3",', issues: ['TEST-323', 'TEST2-323'], shortLink: 'testBlackbox.js' , type: 'Blackbox' })
const unitTestTable = new TestTable().add([unitTest]);
const grayboxTestTable = new TestTable().add([grayboxTest]);
const blackboxTestTable = new TestTable().add([blackboxTest]);

describe("constructor", () => {
    it('constructs from json', () => {
        const map = new TableMap({
            "tables": [{
              "id": "1000000",
              "tests": [
                {
                  "id": "1000000",
                  "name": "A",
                  "link": "../testUnit.js#L3",
                  "issues": [
                    "TEST-1001",
                    "TEST-1002"
                  ],
                  "shortLink": "testUnit.js",
                  "type": "Unit"
                }
              ]
            }]
        });

        expect(map.tables).toMatchSnapshot();
    });
});

describe('add', () => {
    it('adds test to existing table', () => {
        const map = new TableMap({
            "tables": [{
              "id": "1",
              "tests": []
            }]
        });
        expect(map.add(unitTest).tables).toMatchSnapshot();
    });

    it('adds test to new table', () => {
        const map = new TableMap();
        expect(map.add(grayboxTest).tables).toMatchSnapshot();
    });

    it('adds single table', () => {
        const map = new TableMap();
        expect(map.add(unitTestTable).tables).toMatchSnapshot();
    });

    it('adds multiple tests to existing table', () => {
        const map = new TableMap({
            "tables": [{
              "id": "1",
              "tests": []
            }]
        });
        expect(map.add([unitTest, unitTest, unitTest]).tables).toMatchSnapshot();
    });

    it('adds multiple tests to new table', () => {
        const map = new TableMap();
        expect(map.add([grayboxTest, grayboxTest, grayboxTest]).tables).toMatchSnapshot();
    });

    it('adds multiple tests to multiple new tables', () => {
        const map = new TableMap();
        expect(map.add([unitTest, grayboxTest, blackboxTest]).tables).toMatchSnapshot();
    });

    it('adds multiple tables', () => {
        const map = new TableMap();
        expect(map.add([unitTestTable, grayboxTestTable, blackboxTestTable]).tables).toMatchSnapshot();
    });
})

describe('getTests', () => {
    it('returns flattened list of tests', () => {
        const map = new TableMap().add([unitTestTable, grayboxTestTable, blackboxTestTable])
        expect(map.getTests()).toMatchSnapshot();
    })
})

describe('getTestsSplitByIssue', () => {
    it('splits tests that have combined issues into separate tests', () => {
        const map = new TableMap().add([unitTestTable, grayboxTestTable, blackboxTestTable])
        expect(map.getTestsSplitByIssue()).toMatchSnapshot();
    })
})

describe('getTables', () => {
    it('returns tables', () => {
        const map = new TableMap().add([unitTestTable, grayboxTestTable, blackboxTestTable])
        expect(map.getTables()).toEqual([unitTestTable, grayboxTestTable, blackboxTestTable]);
    })
})

describe('sort', () => {
    ['issue', 'id', 'link', 'name', 'type'].forEach(key => {
        it(`sorts each table by ${key}`, () => {
            const table = new TestTable(1).add([unitTest, grayboxTest, blackboxTest]);
            const map = new TableMap().add([table, new TestTable().clone(table).setId(2), new TestTable().clone(table).setId(3)])
            expect(map.sort(key).tables).toMatchSnapshot();
        })
    });

    it('sorts list of tables in ascending id order', () => {
        const table = new TestTable(1).add([unitTest, grayboxTest, blackboxTest]);
        const map = new TableMap().add([table, new TestTable().clone(table).setId(2), new TestTable().clone(table).setId(3)])
        expect(map.sort(undefined, undefined, 1).tables).toMatchSnapshot();
    });

    it('sorts list of tables in descending id order', () => {
        const table = new TestTable().add([unitTest, grayboxTest, blackboxTest]);
        const map = new TableMap().add([table, new TestTable().clone(table).setId(2), new TestTable().clone(table).setId(3)])
        expect(map.sort(undefined, undefined, -1).tables).toMatchSnapshot();
    });
})

describe('count', () => {
    it('returns table count', () => {
        const map = new TableMap().add([unitTestTable, grayboxTestTable, blackboxTestTable])
        expect(map.count()).toEqual(3);
    })
})