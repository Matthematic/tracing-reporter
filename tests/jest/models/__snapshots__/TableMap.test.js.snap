// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add adds multiple tables 1`] = `
Array [
  Table {
    "id": undefined,
    "tests": Array [
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
  Table {
    "id": undefined,
    "tests": Array [
      Test {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
    ],
  },
  Table {
    "id": undefined,
    "tests": Array [
      Test {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
    ],
  },
]
`;

exports[`add adds multiple tests to existing table 1`] = `
Array [
  Table {
    "id": "1",
    "tests": Array [
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
]
`;

exports[`add adds multiple tests to multiple new tables 1`] = `
Array [
  Table {
    "id": "1",
    "tests": Array [
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
  Table {
    "id": "2",
    "tests": Array [
      Test {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
    ],
  },
  Table {
    "id": "3",
    "tests": Array [
      Test {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
    ],
  },
]
`;

exports[`add adds multiple tests to new table 1`] = `
Array [
  Table {
    "id": "2",
    "tests": Array [
      Test {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Test {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Test {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
    ],
  },
]
`;

exports[`add adds single table 1`] = `
Array [
  Table {
    "id": undefined,
    "tests": Array [
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
]
`;

exports[`add adds test to existing table 1`] = `
Array [
  Table {
    "id": "1",
    "tests": Array [
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
]
`;

exports[`add adds test to new table 1`] = `
Array [
  Table {
    "id": "2",
    "tests": Array [
      Test {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
    ],
  },
]
`;

exports[`constructor constructs from json 1`] = `
Array [
  Table {
    "id": "1000000",
    "tests": Array [
      Test {
        "id": "1000000",
        "issues": Array [
          "TEST-1001",
          "TEST-1002",
        ],
        "link": "../testUnit.js#L3",
        "name": "A",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
]
`;

exports[`getTests returns flattened list of tests 1`] = `
Array [
  Test {
    "id": "1",
    "issues": Array [
      "TEST-123",
      "TEST2-123",
    ],
    "link": "\\"../testUnit.js#L3\\",",
    "name": "Unit Test",
    "shortLink": "testUnit.js",
    "type": "Unit",
  },
  Test {
    "id": "2",
    "issues": Array [
      "TEST-223",
      "TEST2-223",
    ],
    "link": "\\"../testGraybox.js#L3\\",",
    "name": "Graybox Test",
    "shortLink": "testGraybox.js",
    "type": "Graybox",
  },
  Test {
    "id": "3",
    "issues": Array [
      "TEST-323",
      "TEST2-323",
    ],
    "link": "\\"../testBlackbox.js#L3\\",",
    "name": "Blackbox Test",
    "shortLink": "testBlackbox.js",
    "type": "Blackbox",
  },
]
`;

exports[`getTestsSplitByIssue splits tests that have combined issues into separate tests 1`] = `
Array [
  Test {
    "id": "1",
    "issues": "TEST-123",
    "link": "\\"../testUnit.js#L3\\",",
    "name": "Unit Test",
    "shortLink": "testUnit.js",
    "type": "Unit",
  },
  Test {
    "id": "1",
    "issues": "TEST2-123",
    "link": "\\"../testUnit.js#L3\\",",
    "name": "Unit Test",
    "shortLink": "testUnit.js",
    "type": "Unit",
  },
  Test {
    "id": "2",
    "issues": "TEST-223",
    "link": "\\"../testGraybox.js#L3\\",",
    "name": "Graybox Test",
    "shortLink": "testGraybox.js",
    "type": "Graybox",
  },
  Test {
    "id": "2",
    "issues": "TEST2-223",
    "link": "\\"../testGraybox.js#L3\\",",
    "name": "Graybox Test",
    "shortLink": "testGraybox.js",
    "type": "Graybox",
  },
  Test {
    "id": "3",
    "issues": "TEST-323",
    "link": "\\"../testBlackbox.js#L3\\",",
    "name": "Blackbox Test",
    "shortLink": "testBlackbox.js",
    "type": "Blackbox",
  },
  Test {
    "id": "3",
    "issues": "TEST2-323",
    "link": "\\"../testBlackbox.js#L3\\",",
    "name": "Blackbox Test",
    "shortLink": "testBlackbox.js",
    "type": "Blackbox",
  },
]
`;

exports[`sort sorts each table by id 1`] = `
Array [
  Table {
    "id": 1,
    "tests": Array [
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
      Test {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Test {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
    ],
  },
  Table {
    "id": 2,
    "tests": Array [
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
    ],
  },
  Table {
    "id": 3,
    "tests": Array [
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
    ],
  },
]
`;

exports[`sort sorts each table by issue 1`] = `
Array [
  Table {
    "id": 1,
    "tests": Array [
      Test {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
      Test {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
  Table {
    "id": 2,
    "tests": Array [
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
  Table {
    "id": 3,
    "tests": Array [
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
]
`;

exports[`sort sorts each table by link 1`] = `
Array [
  Table {
    "id": 1,
    "tests": Array [
      Test {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
      Test {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
  Table {
    "id": 2,
    "tests": Array [
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
  Table {
    "id": 3,
    "tests": Array [
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
]
`;

exports[`sort sorts each table by name 1`] = `
Array [
  Table {
    "id": 1,
    "tests": Array [
      Test {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
      Test {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
  Table {
    "id": 2,
    "tests": Array [
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
  Table {
    "id": 3,
    "tests": Array [
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
]
`;

exports[`sort sorts each table by type 1`] = `
Array [
  Table {
    "id": 1,
    "tests": Array [
      Test {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
      Test {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
  Table {
    "id": 2,
    "tests": Array [
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
  Table {
    "id": 3,
    "tests": Array [
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
    ],
  },
]
`;

exports[`sort sorts list of tables in ascending id order 1`] = `
Array [
  Table {
    "id": 1,
    "tests": Array [
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
      Test {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Test {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
    ],
  },
  Table {
    "id": 2,
    "tests": Array [
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
    ],
  },
  Table {
    "id": 3,
    "tests": Array [
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
    ],
  },
]
`;

exports[`sort sorts list of tables in descending id order 1`] = `
Array [
  Table {
    "id": 3,
    "tests": Array [
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
    ],
  },
  Table {
    "id": 2,
    "tests": Array [
      Object {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
      Object {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Object {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
    ],
  },
  Table {
    "id": 1,
    "tests": Array [
      Test {
        "id": "1",
        "issues": Array [
          "TEST-123",
          "TEST2-123",
        ],
        "link": "\\"../testUnit.js#L3\\",",
        "name": "Unit Test",
        "shortLink": "testUnit.js",
        "type": "Unit",
      },
      Test {
        "id": "2",
        "issues": Array [
          "TEST-223",
          "TEST2-223",
        ],
        "link": "\\"../testGraybox.js#L3\\",",
        "name": "Graybox Test",
        "shortLink": "testGraybox.js",
        "type": "Graybox",
      },
      Test {
        "id": "3",
        "issues": Array [
          "TEST-323",
          "TEST2-323",
        ],
        "link": "\\"../testBlackbox.js#L3\\",",
        "name": "Blackbox Test",
        "shortLink": "testBlackbox.js",
        "type": "Blackbox",
      },
    ],
  },
]
`;
